Qt: Implement Wii settings pane

grimpunch opened the pull request:
![](https://puu.sh/wPdwV/ca12b903a4.png)
Current settings menu is on the left, menu with my additional settings on the right.
I ported over the volume slider for the wiimote speaker and the checkbox toggle for enabling the rumble feature on wiimotes.

leoetlino commented:
These should be in the Wii configuration section, not Controllers configuration, because they cannot be changed at runtime and most Wii settings belong to the Wii config technically.

JosJuice commented:
Placing it the way that this PR does also has the problem of making it look like these settings won't affect anything when Bluetooth passthrough is used.

As a more minor note, we use "Wii Remote" instead of "Wiimote" in all user-facing strings.

grimpunch commented:
Noted, I'll move the options to the right config screen and update the strings.

JosJuice commented:
@dolphin-emu-bot rebuild

grimpunch commented:
I've updated the Settings menu to have a panel for Wii settings:
![](https://puu.sh/wPgJ1/aa0f5a4fa9.png)

I moved the controls there and the panel can be updated to have the rest of the settings from WX.

Starsam80 commented on Source/Core/DolphinQt2/Config/SettingsWindow.cpp:
```diff
@@ -69,6 +70,12 @@ void SettingsWindow::SetupSettingsWidget()
   m_settings_outer->addWidget(audio_pane);
 
   m_settings_outer->addWidget(new PathPane);
+  auto* wii_pane = new WiiPane;
```
This should go after Audio but before Paths to match WX

Starsam80 commented on Source/Core/DolphinQt2/Config/SettingsWindow.cpp:
```diff
@@ -95,6 +111,8 @@ void SettingsWindow::MakeCategoryList()
   AddCategoryToList(tr("Interface"), "browse");
   AddCategoryToList(tr("Audio"), "play");
   AddCategoryToList(tr("Paths"), "browse");
+  AddCategoryToList(tr("Wii"), "Platform_Wii");
```
You should just use `wiimote` instead of `Platform_Wii` and get rid of the extra logic above

Starsam80 commented on Source/Core/DolphinQt2/Settings/WiiPane.cpp:
```diff
@@ -0,0 +1,77 @@
+  auto* basic_group_layout = new QVBoxLayout;
+  basic_group->setLayout(basic_group_layout);
+  m_main_layout->addWidget(basic_group);
+  m_wiimote_motor = new QCheckBox(tr("Enable Wiimote Motor"));
```
As @​JosJuice said:
> As a more minor note, we use "Wii Remote" instead of "Wiimote" in all user-facing strings.

ligfx replied:
This should also use "Rumble" instead of "Motor" in order to match DolphinWX (which has "Wii Remote Rumble").

Starsam80 commented on Source/Core/DolphinQt2/Settings/WiiPane.cpp:
```diff
@@ -0,0 +1,77 @@
+  m_main_layout->addWidget(basic_group);
+  m_wiimote_motor = new QCheckBox(tr("Enable Wiimote Motor"));
+  // Speaker Volume Slider
+  m_wiimote_speaker_volume_label = new QLabel(tr("Wiimote Speaker Volume"));
```
ditto

Starsam80 commented:
@dolphin-emu-bot rebuild

spycrab commented:
I suggest renaming this to ``Qt: Implement Wii settings tab`` or something similar and implementing the rest of the settings from Wx (as you have already stated).

ligfx commented on Source/Core/DolphinQt2/Settings/WiiPane.cpp:
```diff
@@ -0,0 +1,77 @@
+
+void WiiPane::CreateMisc()
+{
+  auto* basic_group = new QGroupBox(tr("Misc Settings"));
```
This group's name should be "Wii Remote Settings," to match WX.

ligfx commented on Source/Core/DolphinQt2/Settings/WiiPane.cpp:
```diff
@@ -0,0 +1,77 @@
+  // Speaker Volume Slider
+  m_wiimote_speaker_volume_label = new QLabel(tr("Wiimote Speaker Volume"));
+  m_wiimote_speaker_volume = new QSlider(Qt::Horizontal);
+  m_wiimote_speaker_volume->setFocusPolicy(Qt::StrongFocus);
```
Why the explicit focus policy?

ligfx commented on Source/Core/DolphinQt2/Settings/WiiPane.cpp:
```diff
@@ -0,0 +1,77 @@
+  m_wiimote_speaker_volume = new QSlider(Qt::Horizontal);
+  m_wiimote_speaker_volume->setFocusPolicy(Qt::StrongFocus);
+  m_wiimote_speaker_volume->setTickPosition(QSlider::TicksBothSides);
+  m_wiimote_speaker_volume->setTickInterval(8);
```
Slider ticks aren't used anywhere else in DolphinQt2, or in the original DolphinWX UI. If the slider value needs to be communicated more accurately than eyeballing its position between "Min" and "Max" (which, yeah, it probably should be, @MayImilae ?), the pattern used elsewhere is to attach a dynamic label afterwards. The audio stretching buffer size slider is a good example of this pattern:

<img width="286" alt="screen shot 2017-07-21 at 3 14 36 pm" src="https://user-images.githubusercontent.com/594093/28484460-526a11d6-6e27-11e7-8bc6-5e53745c7aa4.png">

grimpunch replied:
Noted, i'll change my implementation to do something like this as soon as possible.

leoetlino replied:
For the reference, the official HOME menu uses 10 ticks.

![image](https://user-images.githubusercontent.com/4209061/28487813-26f5259a-6ecc-11e7-98e4-b92d2eaacb82.png)

ligfx commented on Source/Core/DolphinQt2/Settings/WiiPane.cpp:
```diff
@@ -0,0 +1,77 @@
+  m_wiimote_speaker_volume->setFocusPolicy(Qt::StrongFocus);
+  m_wiimote_speaker_volume->setTickPosition(QSlider::TicksBothSides);
+  m_wiimote_speaker_volume->setTickInterval(8);
+  m_wiimote_speaker_volume->setSingleStep(1);
```
This is the default and doesn't need to be specified.

ligfx commented on Source/Core/DolphinQt2/Settings/WiiPane.cpp:
```diff
@@ -0,0 +1,77 @@
+  m_wiimote_speaker_volume->setMaximum(127);
+
+  basic_group_layout->addWidget(m_wiimote_speaker_volume_label);
+  basic_group_layout->addWidget(m_wiimote_speaker_volume,1, Qt::AlignRight);
```
Using a QVBoxLayout doesn't quite accomplish what we want here: even though the slider is being aligned right, it gets shunted onto a separate line.

A pattern used elsewhere for this type of UI is to use a QGridLayout:

```c++
auto* basic_group_layout = new QGridLayout();
basic_group_layout->addWidget(m_wiimote_speaker_volume_label, 0, 0);
basic_group_layout->addWidget(m_wiimote_speaker_volume, 0, 1, 1, -1);
basic_group_layout->addWidget(m_wiimote_motor, 1, 0, 1, -1);
```

You could also make it work by adding the widgets on that row into a QHBoxLayout, and then adding that into the basic group layout.

grimpunch replied:
I agree, your implementation leads to a much more aesthetically pleasing layout.

![wiiremotesettings](https://user-images.githubusercontent.com/1707054/28528601-16c0cb00-7086-11e7-8dcb-3dcd15d4894f.PNG)

ligfx commented on Source/Core/DolphinQt2/Settings/WiiPane.h:
```diff
@@ -0,0 +1,35 @@
+#include <QWidget>
+
+class QCheckBox;
+class QComboBox;
```
This forward-declaration isn't needed.

ligfx commented on Source/Core/DolphinQt2/Config/SettingsWindow.cpp:
```diff
@@ -69,6 +70,12 @@ void SettingsWindow::SetupSettingsWidget()
+  auto* wii_pane = new WiiPane;
+  connect(this, &SettingsWindow::EmulationStarted,
+          [wii_pane] { wii_pane->OnEmulationStateChanged(true); });
+  connect(this, &SettingsWindow::EmulationStopped,
```
Please combine this into the other emulation state callbacks above.

spycrab commented:
@dolphin-emu-bot rebuild

grimpunch commented:
Working on implementing the rest of the menu, here's how everything looks presently with the rest of the menu's features ported from Wx.

![](https://puu.sh/wRXB9/e0c5e94d7e.png)
I'm hoping to have all the loading/saving and implemented tomorrow.

grimpunch commented:
Rebased on top of current changes and resolved the merge conflict.
Continuing to integrate the other functionality now.

spycrab commented:
@dolphin-emu-bot rebuild

grimpunch commented:
Rebased this all on one commit, still hopeful someone knows whats up with the remove code not working.

grimpunch commented:
I've fixed the issue where removal was not happening , the cause was improperly casting the device name when saving usb passthrough devices. 
Ready for build/review.

spycrab commented:
@dolphin-emu-bot rebuild

Sevenanths commented:
Why would there need to be a panel with icons? When the icons don't really _add_ something to the user experience, tab pages should work just fine. You're just wasting vertical space right now.

spycrab commented:
> Why would there need to be a panel with icons? When the icons don't really add something to the user experience, tab pages should work just fine. You're just wasting vertical space right now.

Not relevant to this PR.

spycrab commented:
@dolphin-emu-bot rebuild

Sevenanths commented:
My apologies! I can't read.

spycrab commented:
Anything else to do here? @lioncash @ligfx

ligfx commented on Source/Core/DolphinQt2/Settings/USBDeviceAddToWhitelistDialog.cpp:
```diff
@@ -0,0 +1,169 @@
+  const auto& current_devices = USBUtils::GetInsertedDevices();
+  if (current_devices == m_shown_devices)
+    return;
+  QSignalBlocker blocker(usb_inserted_devices_list);  // unblocks when out of scope.
```
Blocking signals seems heavy-handed.

grimpunch replied:
You're right, I don't think it is needed at all.

ligfx commented on Source/Core/DolphinQt2/Settings/USBDeviceAddToWhitelistDialog.cpp:
```diff
@@ -0,0 +1,169 @@
+    // i18n: Here, VID means Vendor ID (for a USB device).
+    QErrorMessage* error = new QErrorMessage();
+    error->setWindowTitle(tr("USB Whitelist Error"));
+    error->showMessage(tr("The entered VID is invalid."));
```
Using `QMessageBox::warning` probably makes more sense for these. QErrorMessage has a "Show this message again" checkbox that lets the user disable future dialogs (which also won't actually do anything, because it's being instantiated over and over again).

ligfx commented on Source/Core/DolphinQt2/Settings/USBDeviceAddToWhitelistDialog.cpp:
```diff
@@ -0,0 +1,169 @@
+  SConfig::GetInstance().m_usb_passthrough_devices.emplace(vid, pid);
+  SConfig::GetInstance().SaveSettings();
+  // Not the nicest way of doing this but...
+  WiiPane* p = static_cast<WiiPane*>(parent());
```
Instead, `WiiPane` should call `PopulateUSBPassthroughListWidget()` on itself after waiting for this dialog to close.

grimpunch replied:
Working my way through these now, and i've done this (and made the function private too).

ligfx commented on Source/Core/DolphinQt2/Settings/USBDeviceAddToWhitelistDialog.cpp:
```diff
@@ -0,0 +1,169 @@
+  close();
+}
+
+void USBDeviceAddToWhitelistDialog::OnRefreshDevicesTimer()
```
There's no need for this function: just reference RefreshDeviceList directly.

ligfx commented on Source/Core/DolphinQt2/Settings/USBDeviceAddToWhitelistDialog.cpp:
```diff
@@ -0,0 +1,169 @@
+void USBDeviceAddToWhitelistDialog::OnDeviceSelection()
+{
+  // Not the nicest way of doing this but...
+  QString device = usb_inserted_devices_list->currentItem()->text().left(9);
```
See my comment on the device list in WiiPane on how to do this in a non-fragile way using the data field of QListWidgetItems.

grimpunch commented:
Okay I will look into these during the weekend or later this week in general if possible

grimpunch commented on Source/Core/DolphinQt2/Settings/USBDeviceAddToWhitelistDialog.cpp:
```diff
@@ -0,0 +1,169 @@
+  enter_device_id_label->setAlignment(Qt::AlignCenter);
+  main_layout->addWidget(enter_device_id_label);
+
+  entry_form_layout = new QFormLayout();
```
Thanks, looks identical, and a lot simpler.

grimpunch commented on Source/Core/DolphinQt2/Settings/WiiPane.cpp:
```diff
@@ -0,0 +1,277 @@
+void WiiPane::OnUSBWhitelistAddButton()
+{
+  USBDeviceAddToWhitelistDialog* usb_whitelist_dialog = new USBDeviceAddToWhitelistDialog(this);
+  usb_whitelist_dialog->show();
```
I agree, made it so.

grimpunch commented on Source/Core/DolphinQt2/Settings/WiiPane.cpp:
```diff
@@ -0,0 +1,277 @@
+{
+  auto* whitelisted_usb_passthrough_devices_group =
+      new QGroupBox(tr("Whitelisted USB Passthrough Devices"));
+  auto* whitelisted_usb_passthrough_devices_v_layout = new QVBoxLayout();
```
Changed the layout to work like this now.

grimpunch commented on Source/Core/DolphinQt2/Settings/WiiPane.cpp:
```diff
@@ -0,0 +1,277 @@
+void WiiPane::OnUSBWhitelistRemoveButton()
+{
+  std::set<std::pair<u16, u16>> test_set = SConfig::GetInstance().m_usb_passthrough_devices;
+  QString device = m_whitelist_usb_list->currentItem()->text().left(9);
```
Not really sure how to do that either, so @spycrab , your help would be appreciated!
Thanks.

grimpunch commented:
@ligfx I've committed changes to everything addressing nearly all issues above.
+1

spycrab commented:
@dolphin-emu-bot rebuild

grimpunch commented:
@spycrab @ligfx  I think the linter is broken again, please let me know if there are any more issues with this PR.
Thanks.

JosJuice commented:
In order to avoid that problem with the lint bot, you have to rebase your branch on the latest version of master.

spycrab commented:
@grimpunch Any updates on this?

Helios747 commented:
Can you rebase this and fix lint?

grimpunch commented:
@spycrab Hi, sorry for late update,
I'm now employed in a position where I could not contribute to the project due to my access to dev hardware, so I have not been updating my work for that reason, sorry! 
Feel free to cherry pick the relevant code out that may be used for the settings pane or someone else can propose a pull request for this issue.

spycrab commented:
@grimpunch Thanks for your reply!

I took the liberty of rebasing and updating your code and creating a new PR.

While it's unfortunate that you'll be unable to contribute for now, I wish you the best of luck with your job!

JosJuice commented:
Merged as PR #6336.
